src/wasm.h

```diff
@@ -1233,13 +1234,104 @@ std::ostream& Expression::print(std::ostream &o, unsigned indent) {
   return o;
 }
 
+template<typename SubType, typename ReturnType>
+struct WasmWalkerBase : public WasmVisitor<SubType, ReturnType> {
+  virtual void walk(Expression*& curr) { abort(); }
+  virtual void startWalk(Function *func) { abort(); }
+  virtual void startWalk(Module *module) { abort(); }
+};
+
+template<typename ParentType>
+struct ChildWalker : public WasmWalkerBase<ChildWalker<ParentType>, void> {
+  ParentType& parent;
+
+  ChildWalker(ParentType& parent) : parent(parent) {}
+
+  void visitBlock(Block *curr) {
+    ExpressionList& list = curr->list;
+    for (size_t z = 0; z < list.size(); z++) {
```

Comment:
It depends: some datastructures don't inline well, and sometimes the calls in the loop confuse the optimizer into emitting the `size` call on every iteration.

Range-based-for will indeed fail if there's mutation, would be good to validate.